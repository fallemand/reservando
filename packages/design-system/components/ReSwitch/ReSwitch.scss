@import "../../styles/variables";
$duration: 0.3s;

.re-switch {
  display: flex;
  align-items: center;

  &__input {
    opacity: 0;
    width: 0;
    height: 0;
    margin: 0;
  }

  &__label {
    margin-left: 0.5em;
  }

  &__control {
    font-size: inherit;
    position: relative;
    display: inline-block;
    width: 3em;
    height: 1.5em;
    line-height: calc(1em - 2px);
    padding: 0;
    vertical-align: middle;
    border-radius: 1em;
    background-color: $grey-400;
    border: 0;
    cursor: pointer;
    transition: all $duration cubic-bezier(0.35, 0, 0.25, 1);
    box-shadow: inset 0px 1px 2px rgba(0, 50, 77, 0.2);

    &::after {
      position: absolute;
      width: calc(1.5em - 4px);
      height: calc(1.5em - 4px);
      left: 3px;
      top: 2px;
      border-radius: 50% 50%;
      background-color: $white;
      content: " ";
      cursor: pointer;
      transform: scale(1);
      transition: left $duration cubic-bezier(0.35, 0, 0.25, 1);
      animation-timing-function: cubic-bezier(0.35, 0, 0.25, 1);
      animation-duration: $duration;
      animation-name: switchOff;
      box-shadow: 0px 0px 1px rgba(0, 50, 77, 0.1), 0px 0px 2px rgba(0, 50, 77, 0.13),
        0px 1px 4px rgba(0, 50, 77, 0.1);
    }

    &:hover::after {
      transform: scale(1.1);
      animation-name: switchOn;
    }

    &:focus {
      box-shadow: 0 0 0 2px tint($primary-700 80%);
      outline: none;
    }

    &--checked {
      background-color: $primary-600;

      &::after {
        left: 1.5em;
      }
    }

    &--disabled {
      cursor: no-drop;

      &::after {
        animation-name: none;
        cursor: no-drop;
      }

      &:hover::after {
        transform: scale(1);
        animation-name: none;
      }
    }
  }
}

@keyframes switchOn {
  0% {
    transform: scale(1);
  }
  50% {
    transform: scale(1.25);
  }
  100% {
    transform: scale(1.1);
  }
}

@keyframes switchOff {
  0% {
    transform: scale(1.1);
  }
  100% {
    transform: scale(1);
  }
}
