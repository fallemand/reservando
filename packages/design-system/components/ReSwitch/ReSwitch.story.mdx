import { Meta, Story, ArgsTable, Canvas, Description } from '@storybook/addon-docs/blocks';
import { ReSwitch } from "@reservando/design-system";

<Meta title="Components/Switch" component={ReSwitch} />

# ReSwitch

<Canvas>
  <Story name="basic">
    {{
      components: { ReSwitch },
      data() {
        return {
          checked: true,
        };
      },
      template: `
        <div>
          <ReSwitch v-model="checked" id="switch" label="Label" />
        </div>
      `,
    }}
  </Story>
</Canvas>

<Description of={ReSwitch} />

## How to use

```html
<template>
  <ReSwitch v-model="value" id="my-switch" label="Label" />
</template>

<script lang="ts">
  import { ReSwitch } from '@reservando/design-system';
</script>
```

<ArgsTable of={ReSwitch} />

## Examples

#### Different states

<Canvas>
  <Story name="disabled">
    {{
      components: { ReSwitch },
      template: `
        <div>
          <ReSwitch id="disabled" disabled class="mb-2" label="Disabled" />
          <ReSwitch id="disabled-checked" class="mb-2" :modelValue="true" disabled label="Disabled" />
        </div>
      `,
    }}
  </Story>
</Canvas>
