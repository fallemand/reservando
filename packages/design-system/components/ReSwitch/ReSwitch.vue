<template>
  <div class="re-switch" :class="$attrs.class">
    <input
      id="id"
      ref="input"
      type="checkbox"
      class="re-switch__input"
      role="switch"
      v-bind="{
        ...$attrs,
        class: undefined,
        onChange: undefined,
      }"
      :aria-checked="`${modelValue}`"
      :checked="modelValue"
      aria-label="checkbox"
    />
    <button
      type="button"
      class="re-switch__control"
      :disabled="disabled"
      :class="{
        're-switch__control--checked': modelValue,
        're-switch__control--disabled': disabled,
      }"
      @click="emitChange"
    ></button>
    <label v-if="label" class="re-switch__label" :for="id">{{ label }}</label>
  </div>
</template>

<script lang="ts">
import { defineComponent } from "vue";

const ReSwitch = defineComponent({
  name: "ReSwitch",
  inheritAttrs: false,
  props: {
    modelValue: {
      type: Boolean,
      required: true,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    label: { type: String, default: null },
    id: { type: String, required: true },
  },
  methods: {
    emitChange(): void {
      this.$emit("update:modelValue", !this.modelValue);
    },
  },
});

export default ReSwitch;
</script>

<style lang="scss" src="./ReSwitch.scss" />
