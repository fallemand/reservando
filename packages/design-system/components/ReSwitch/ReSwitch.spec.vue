<script>
import { shallowMount } from "@vue/test-utils";
import ReSwitch from "./ReSwitch.vue";

describe("ReSwitch component", () => {
  const props = {
    id: "SWITCH-ID",
    modelValue: false,
  };

  it("should render with default props", () => {
    const wrapper = shallowMount(ReSwitch, { props });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should toggle switch state when clicked", async () => {
    const wrapper = shallowMount(ReSwitch, {
      props,
    });
    const control = wrapper.find(".re-switch__control");
    await control.trigger("click");

    expect(wrapper.emitted()["update:modelValue"][0]).toEqual([true]);
  });

  it("should not trigger if disabled", async () => {
    const wrapper = shallowMount(ReSwitch, {
      props: {
        ...props,
        disabled: true,
      },
    });
    const control = wrapper.find(".re-switch__control");
    await control.trigger("click");

    expect(wrapper.emitted()["update:modelValue"]).toEqual(undefined);
  });
});
</script>
