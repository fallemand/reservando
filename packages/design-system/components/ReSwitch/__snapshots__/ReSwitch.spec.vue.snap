// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReSwitch component should render with default props 1`] = `
<div
  class="re-switch"
>
  <input
    aria-checked="false"
    aria-label="checkbox"
    class="re-switch__input"
    id="SWITCH-ID"
    role="switch"
    type="checkbox"
  />
  <button
    class="re-switch__control"
    type="button"
  />
  <!--v-if-->
</div>
`;
